core/src/main/java/tc/oc/pgm/action/replacements/ReplacementParser.java

```diff
@@ -0,0 +1,193 @@
+    scope = parseScope(el, scope);
+    Formula<T> formula = parser.formula(scope, el, "value").orNull();
+    var fallback = parser.formattedText(el, "fallback").child().optional(empty());
+    var children = el.getChildren("case");
+    var branches = new ArrayList<SwitchReplacement.Branch>(children.size());
+
+    for (var innerEl : children) {
+      var filter = parser.filter(innerEl, "filter").orNull();
+      var valueRange =
+          XMLUtils.parseNumericRange(Node.fromChildOrAttr(innerEl, "match"), Double.class, null);
+      if (filter == null && valueRange == null) {
+        throw new InvalidXMLException(
+            "At least a filter or a match attribute must be specified", innerEl);
+      }
+
+      if (valueRange != null && formula == null) {
+        throw new InvalidXMLException(
+            "A match attribute is specified but there's no switch value to bind to", innerEl);
+      }
+
+      var result = parser.formattedText(innerEl, "result").required();
+
+      if (filter != null) {
+        Class<T> finalScope = scope;
+        features.validate(
+            filter,
+            (f, n) -> {
+              if (!f.respondsTo(finalScope)) {
+                throw new InvalidXMLException(
+                    "Expected a filter that can respond to "
+                        + finalScope.getSimpleName()
+                        + ". The filter "
+                        + f.getDefinitionType().getSimpleName()
+                        + " only responds to "
+                        + Filterables.scope(f).getSimpleName()
+                        + " or lower",
+                    n);
+              }
+            },
+            Node.fromChildOrAttr(innerEl, "filter"));
+      }
+
+      branches.add(new SwitchReplacement.Branch(
+          result,
+          valueRange == null ? Range.all() : valueRange,
+          filter == null ? StaticFilter.ALLOW : filter));
+    }
+
+    return new SwitchReplacement<>(scope, branches, fallback, formula);
+  }
+
+  @MethodParser("replacement")
+  public <T extends Filterable<?>> Replacement<T> parseReference(Element el, Class<T> scope)
+      throws InvalidXMLException {
+    if (isTopLevel)
+      throw new InvalidXMLException(
+          "References to replacements at the root level are not allowed", el);
+
+    @SuppressWarnings("unchecked")
+    Replacement<T> replacement = (Replacement<T>) features.resolve(new Node(el), Replacement.class);
+
+    validateScope(replacement, scope, el);
+    return replacement;
+  }
+
+  private void validateScope(Replacement<?> definition, Class<?> scope, Element el)
+      throws InvalidXMLException {
+    Class<?> definitionScope = definition.getScope();
+    if (!definitionScope.isAssignableFrom(scope))
+      throw new InvalidXMLException(
+          "Wrong replacement scope, got "
+              + definitionScope.getSimpleName()
+              + " but expected "
+              + scope.getSimpleName(),
+          el);
+  }
```

Comment:
As mentioned before, relegate this to a method within the replacement; let it decide what the logic for rejecting a scope should be